src/panels/lovelace/components/hui-yaml-editor.ts

```diff
@@ -0,0 +1,98 @@
+// @ts-ignore
+import CodeMirror from "codemirror";
+import "codemirror/mode/yaml/yaml";
+// @ts-ignore
+import codeMirrorCSS from "codemirror/lib/codemirror.css";
+import { fireEvent } from "../../../common/dom/fire_event";
+declare global {
+  interface HASSDomEvents {
+    "yaml-changed": {
+      value: string;
+    };
+    "yaml-save": {};
+  }
+}
+
+export class HuiYamlEditor extends HTMLElement {
+  public codemirror: CodeMirror;
+  private _value: string;
+
+  public constructor() {
+    super();
+    this._value = "";
+    const shadowRoot = this.attachShadow({ mode: "open" });
+    shadowRoot.innerHTML = `
+            <style>
+               ${codeMirrorCSS}
+               .CodeMirror {
+                  height: var(--code-mirror-height, 300px);
+                  direction: var(--code-mirror-direction, ltr);
+                }
+                .CodeMirror-gutters {
+                  border-right: 1px solid var(--paper-input-container-color, var(--secondary-text-color));
+                  background-color: var(--paper-dialog-background-color, var(--primary-background-color));
+                  transition: 0.2s ease border-right;
+                }
+                .CodeMirror-focused .CodeMirror-gutters {
+                  border-right: 2px solid var(--paper-input-container-focus-color, var(--primary-color));;
+                }
+                .CodeMirror-linenumber {
+                  color: var(--paper-dialog-color, var(--primary-text-color));
+                }
+            </style>`;
+  }
+
+  set value(value: string) {
+    if (this.codemirror) {
+      if (value !== this.codemirror.getValue()) {
+        this.codemirror.setValue(value);
+      }
+    }
+    this._value = value;
+  }
+
+  get value(): string {
+    return this.codemirror.getValue();
+  }
+
+  get hasComments(): boolean {
+    return this.shadowRoot!.querySelector("span.cm-comment") ? true : false;
+  }
+
+  public connectedCallback(): void {
+    if (!this.codemirror) {
+      this.codemirror = CodeMirror(this.shadowRoot, {
+        value: this._value,
+        lineNumbers: true,
+        mode: "yaml",
+        tabSize: 2,
+        autofocus: true,
+        extraKeys: {
+          Tab: (cm: CodeMirror) => {
+            const spaces = Array(cm.getOption("indentUnit") + 1).join(" ");
+            cm.replaceSelection(spaces);
+          },
+        },
+      });
+      this.codemirror.commands.save = () => {
+        fireEvent(this, "yaml-save");
+      };
+      fireEvent(this, "yaml-changed", { value: this._value });
```

Comment:
Why do we fire this?

Reply:
It triggers the preview, should actually solve that in the editor...